accepted/2023/wasm-browser-threads.md

```diff
@@ -0,0 +1,268 @@
+# Multi-threading on a browser
+
+## Goals
+ - CPU intensive workloads on dotnet thread pool
+ - enable blocking .Wait APIs from C# user code on all threads
+     - Current public API throws PNSE for it
+     - This is core part on MT value proposition.
+     - If people want to use existing MT code-bases, most of the time, the code is full of locks. 
+     - People want to use existing desktop/server multi-threaded code as is.
+ - allow HTTP and WS C# APIs to be used from any thread
+     - Underlying JS object have thread affinity
+ - JSImport/JSExport interop in maximum possible extent
+ - don't change/break single threaded build. †
+
+## Lower priority goals
+ - try to make it debugging friendly
+ - implement crypto via `subtle` browser API
+ - allow calls to synchronous JSExport from UI thread (callback)
+
+<sub><sup>† Note: all the text below discusses MT build only, unless explicit about ST build.</sup></sub>
+
+## Context - Problems
+**1)** If you have multithreading, any thread might need to block while waiting for any other to release a lock.
+     - locks are in the user code, in nuget packages, in Mono VM itself
+     - there are managed and un-managed locks
+     - in single-threaded build of the runtime, all of this is NOOP. That's why it works on UI thread.
+
+**2)** UI thread in the browser can't synchronously block
+    - you can spin-lock but it's bad idea.
+        - Deadlock: when you spin-block, the JS timer loop and any messages are not pumping. But code in other threads may be waiting for some such event to resolve.
+        - It eats your battery
+        - Browser will kill your tab at random point (Aw, snap).
+        - It's not deterministic and you can't really test your app to prove it harmless.
+    - Firefox (still) has synchronous XHR which could be captured by async code in service worker
+        - it's deprecated legacy API
+        - but other browsers don't and it's unlikely they will implement it
+        - there are deployment and security challenges with it
+    - all the other threads/workers could synchronously block
+    - if we will have managed thread on the UI thread, any `lock` or Mono GC barrier could cause spin-wait
+        - in case of Mono code,  we at least know it's short duration
+        - we should prevent it from blocking in user code
+
+**3)** JavaScript engine APIs and objects have thread affinity. 
+    - The DOM and few other browser APIs are only available on the main UI "thread"
+        - and so, you need to have C# interop with UI, but you can't block there.
+    - HTTP & WS objects have affinity, but we would like to consume them (via Streams) from any managed thread
```

Comment:
note that it may actually be a bad idea to support this. If someone is calling Stream.Read(byte[256]) from the wrong thread repeatedly the performance is going to be really bad to an extent that is not true in any other environment

Reply:
This is requirements from Blazor team.

Reply:
Is Blazor team aware of the fact that satisfying this requirement will come with performance and reliability cliffs?

Reply:
I added @SteveSandersonMS to review here and will add more details.